readers/combined.cc

```diff
@@ -484,25 +486,29 @@ future<mutation_reader_merger::needs_merge> mutation_reader_merger::prepare_one(
             // end are out of data for good for the current range.
             _halted_readers.push_back(rk);
         } else if (_fwd_mr == mutation_reader::forwarding::no) {
-            _to_remove.splice(_to_remove.end(), _all_readers, rk.reader);
-            if (_to_remove.size() >= 4) {
-                auto to_remove = std::move(_to_remove);
-                to_close = parallel_for_each(to_remove, [] (flat_mutation_reader_v2& r) {
+                // FIXME: indentation
+                flat_mutation_reader_v2 r = std::move(*rk.reader);
+                _all_readers.erase(rk.reader);
+                _pending_close++;
+                auto d = defer([this] { _pending_close--; });
+                _to_close = _to_close.then([this, r = std::move(r), d = std::move(d)] () mutable {
```

Comment:
If close returns with an unavailable furure, `d` will be destroyed.
You should capture it in a close.finally continuation

Reply:
even sinpler. just do the deferred action - decrementing the counter, in the finally continuation.
you don't need the deferred_action at all since close doesn't throw.

Reply:
otherwise, looks good.

Reply:
The idea was that _pending_close counts the # of unneeded readers which are waiting for their turn to launch close() call.

So if we're closing r1, and r2, r3 and r4 are waiting for their turn, then _pending_close == 3. As soon as we add r5, then prepare_one() waits for completion of close on r1 .. r5.

But it's an off-by-one. Meaning we could allow up to 5 pending close calls.

IIRC, finally() is very expensive, so we can just do the follow as close() is noexcept qualified:

```
_to_close.then([...]
  return r.close().then([] { decrement counter; });
});
```

Thanks!

Reply:
> IIRC, finally() is very expensive,

It just has an extra try / catch over then_wrapped in this case where only close may fail.

> so we can just do the follow as close() is noexcept qualified:

Although close is noexcept qualified it might still fail, just return that failure as an exceptional future.
You want to decrement the counter in that case too.

Although
```
  return r.close().then([] { decrement counter; });
```
would practically work since close in the read-only case has no reason to fail (if the file was written to ans wasn't flushed before close, close might return an error to write from flush),

We can just batch the bookkeeping too, and simply increment the count here,
and decrement it below in batch, like:
```
    future<> to_close = make_ready_future<>();
    if (_pending_close >= 4) {
        _pending_close = 0;
        to_close = std::exchange(_to_close, make_ready_future<>());
    }
```

Reply:
Why create d at all? Just decrement the counter in the continuation.

Reply:
With my suggestion above (https://github.com/scylladb/scylladb/pull/11167#discussion_r933758505)
there shouldn't be any deferred action.

Reply:
Yes.

Write a short comment - people don't understand and complain.
Write a long comment - people don't read it.